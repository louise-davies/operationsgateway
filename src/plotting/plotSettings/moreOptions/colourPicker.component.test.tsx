import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ColourPicker from './colourPicker.component';

describe('ColourPicker component', () => {
  const props: React.ComponentProps<typeof ColourPicker> = {
    channelName: 'CHANNEL_1',
    colour: '#ff0000',
    changeColour: jest.fn(),
  };

  it('renders coloured square when popover is not open', () => {
    const view = render(<ColourPicker {...props} />);
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders colour picker when square is clicked', async () => {
    const user = userEvent.setup();
    const view = render(<ColourPicker {...props} />);

    await user.click(screen.getByLabelText(`Pick ${props.channelName} colour`));

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders marker colour picker when marker prop is true', async () => {
    const user = userEvent.setup();
    const view = render(<ColourPicker {...props} marker={true} />);

    await user.click(
      screen.getByLabelText(`Pick ${props.channelName} marker colour`)
    );

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('toggles lockColour state when the Same as line switch is clicked', async () => {
    const user = userEvent.setup();
    render(<ColourPicker {...props} marker={true} />);

    await user.click(
      screen.getByLabelText(`Pick ${props.channelName} marker colour`)
    );

    const switchElement = screen.getByRole('checkbox');

    await user.click(switchElement);

    expect(switchElement).toHaveProperty('checked', true);

    await user.click(switchElement);

    expect(switchElement).toHaveProperty('checked', false);
  });

  it('calls changeColour with the same colour when Same as line switch is toggled', async () => {
    const user = userEvent.setup();
    render(<ColourPicker {...props} marker={true} sameAsLine={true} />);

    await user.click(
      screen.getByLabelText(`Pick ${props.channelName} marker colour`)
    );

    const switchElement = screen.getByRole('checkbox');
    await user.click(switchElement);

    expect(props.changeColour).toHaveBeenCalledWith(props.colour);

    await user.click(switchElement);

    expect(props.changeColour).toHaveBeenCalledWith('');
  });

  it('calls changeColour when new colour is picked', async () => {
    const user = userEvent.setup();
    render(<ColourPicker {...props} />);

    await user.click(screen.getByLabelText(`Pick ${props.channelName} colour`));

    await user.click(screen.getByLabelText('Color'));

    expect(props.changeColour).toHaveBeenCalledWith(expect.anything());
  });

  it('closes colour picker only when you click outside of it', async () => {
    const user = userEvent.setup();
    render(
      <div>
        <div>Outside</div>
        <ColourPicker {...props} />
      </div>
    );

    await user.click(screen.getByLabelText(`Pick ${props.channelName} colour`));

    screen.getByLabelText('Color');

    // don't think this could actually occur with a real mouse, but oh well
    fireEvent.mouseDown(
      screen.getByLabelText(`Pick ${props.channelName} colour`)
    );
    fireEvent.click(screen.getByLabelText('Color'));

    screen.getByLabelText('Color');

    await user.click(screen.getByText('Outside'));

    expect(screen.queryByLabelText('Color')).not.toBeInTheDocument();
  });

  it('closes colour picker when you click the square', async () => {
    const user = userEvent.setup();
    render(<ColourPicker {...props} />);

    await user.click(screen.getByLabelText(`Pick ${props.channelName} colour`));

    screen.getByLabelText('Color');

    await user.click(screen.getByLabelText(`Pick ${props.channelName} colour`));

    expect(screen.queryByLabelText('Color')).not.toBeInTheDocument();
  });
});
