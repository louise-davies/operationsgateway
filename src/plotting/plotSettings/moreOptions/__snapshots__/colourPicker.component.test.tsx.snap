// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColourPicker component > renders colour picker when square is clicked 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-79elbk"
  >
    <button
      aria-haspopup="dialog"
      aria-label="Pick CHANNEL_1 colour"
      class="MuiBox-root css-3mx4ey"
    />
    <div
      class="MuiBox-root css-eggkpd"
      role="dialog"
    >
      <div
        class="react-colorful"
      >
        <div
          class="react-colorful__saturation"
          style="background-color: rgb(255, 0, 0);"
        >
          <div
            aria-label="Color"
            aria-valuetext="Saturation 100%, Brightness 100%"
            class="react-colorful__interactive"
            role="slider"
            tabindex="0"
          >
            <div
              class="react-colorful__pointer react-colorful__saturation-pointer"
              style="top: 0%; left: 100%;"
            >
              <div
                class="react-colorful__pointer-fill"
                style="background-color: rgb(255, 0, 0);"
              />
            </div>
          </div>
        </div>
        <div
          class="react-colorful__hue react-colorful__last-control"
        >
          <div
            aria-label="Hue"
            aria-valuemax="360"
            aria-valuemin="0"
            aria-valuenow="0"
            class="react-colorful__interactive"
            role="slider"
            tabindex="0"
          >
            <div
              class="react-colorful__pointer react-colorful__hue-pointer"
              style="top: 50%; left: 0%;"
            >
              <div
                class="react-colorful__pointer-fill"
                style="background-color: rgb(255, 0, 0);"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ColourPicker component > renders coloured square when popover is not open 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-79elbk"
  >
    <button
      aria-haspopup="dialog"
      aria-label="Pick CHANNEL_1 colour"
      class="MuiBox-root css-3mx4ey"
    />
  </div>
</DocumentFragment>
`;
